src/idl/src/annotation.c

```diff
@@ -306,6 +341,85 @@ set_nested(
   return IDL_RETCODE_OK;
 }
 
+static idl_retcode_t
+annotate_default(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *annotation_appl,
+  idl_node_t *node)
+{
+  const idl_literal_t *value;
+  idl_type_spec_t *mem_spec;
+
+  assert(annotation_appl);
+  assert(annotation_appl->parameters);
+
+  if (!idl_is_member(node)) {
+    idl_error(pstate, idl_location(annotation_appl),
+      "@default can only be assigned to member nodes");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+  mem_spec = ((idl_member_t*)node)->type_spec;
+
+  value = annotation_appl->parameters->const_expr;
+  if (!value || !idl_is_literal(value)) {
+    idl_error(pstate, idl_location(annotation_appl),
+      "@default requires a value");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  //check whether type of literal matches type_spec of member
+  switch(idl_mask(value)) {
+    case IDL_LONG:
+      if (!idl_is_integer_type(mem_spec)) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "value type of @default does not match member");
+        return IDL_RETCODE_SEMANTIC_ERROR;
+      }
+      break;
+    case IDL_DOUBLE:
+      if (!idl_is_floating_pt_type(mem_spec)) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "value type of @default does not match member");
+        return IDL_RETCODE_SEMANTIC_ERROR;
+      }
+      break;
+    default:
+    {
+      idl_mask_t value_type = (idl_mask(value) ^ IDL_LITERAL);
+      if (idl_mask(mem_spec) != value_type) {
+        idl_error(pstate, idl_location(annotation_appl),
+          "value type of @default does not match member");
+        return IDL_RETCODE_SEMANTIC_ERROR;
+      }
+    }
+  }
```

Comment:
This too, should have been done by the parser already.

Reply:
I think that there might be a misunderstanding here, this is checking whether the type in the literal (the parameter field of the `@default` annotation) is compatible with that of the type it is being annotated to (so it should reject something like this: `@default('a') long l;`